lib/tree.rb

```diff
@@ -2,61 +2,191 @@ class TreeNode
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    if new_node.key <= current.key
+      if current.left
+        add_helper(current.left, new_node)
+      else
+        current.left = new_node
+      end
+    else
+      if current.right
+        add_helper(current.right, new_node)
+      else
+        current.right = new_node
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(h) if  balanced, where h is the height of the tree
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    current = @root
+    while !current.nil?
+      if key == current.key
+        return current.value
+      elsif key <= current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n^2) where n is the number of nodes
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    stack = []
+    output = []
+    current = root
+    while current || !stack.empty?
+      while !current.nil?
+        stack.push(current)
+        current = current.left
+      end
+      current = stack.pop
+      new_hash = Hash.new
+      new_hash[:key] = current.key
+      new_hash[:value] = current.value
+      output.push(new_hash);  #storing the data
+      current = current.right
+    end
+    return output
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: if well balanced, O(h) where h is the height of the tree
+  # Space Complexity: O(h)
   def preorder
-    raise NotImplementedError
+    stack = []
+    output = []
+
+    current = @root
+
+    return [] if current.nil?
+    stack.push(current)
+
+    while !stack.empty?
+      current = stack.pop()
+      new_hash = Hash.new
+      new_hash[:key] = current.key
+      new_hash[:value] = current.value
+      output.push(new_hash)
+      if current.right
+        stack.push(current.right)
+      end
+      if current.left
+        stack.push(current.left)
+      end
+    end
+    return output
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: if balanced, O(h) where h is the height of the tree
+  # Space Complexity: O(h)
   def postorder
-    raise NotImplementedError
+    stack = []
+    output = []
+    return [] if @root.nil?
+    stack.push(@root)
+    while !stack.empty?
+      current = stack.pop()
+      new_hash = Hash.new
+      new_hash[:key] = current.key
+      new_hash[:value] = current.value
+      output.push(new_hash)
+      if current.left
+        stack.push(current.left)
+      end
+      if current.right
+        stack.push(current.right)
+      end
+    end
+    return output.reverse
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: if balanced, O(h) where h is the height of the tree
```

Comment:
Since you have to visit each node to figure out the height, it's O(n) for time complexity.  Space is different.